File: 0011-container-with-most-water/0011-container-with-most-water.go

func maxArea(height []int) int {
	leftIndex := 0
	rightIndex := len(height) - 1
	maxWater := 0

	for leftIndex < rightIndex {
		currentWidth := rightIndex - leftIndex
		currentHeight := 0

		if height[leftIndex] < height[rightIndex] {
			currentHeight = height[leftIndex]
		} else {
			currentHeight = height[rightIndex]
		}

		currentWater := currentWidth * currentHeight

		if currentWater > maxWater {
			maxWater = currentWater
		}
		if height[leftIndex] < height[rightIndex] {
			leftIndex++
		} else {
			rightIndex--
		}
	}
	return maxWater
}

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    def maxArea(self, height: list[int]) -> int:
        max_water = 0
        left_index = 0
        right_index = len(height) - 1
        
        while left_index < right_index:
            current_width = right_index - left_index
            current_height = min(height[left_index], height[right_index])
            current_water = current_width * current_height
            max_water = max(max_water, current_water)
            if height[left_index] < height[right_index]:
                left_index += 1
            else:
                right_index -= 1
        return max_water
